x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/DataFrameAnalyticsLazyStartIT.java

```diff
@@ -0,0 +1,234 @@
+        return CollectionUtils.appendToCopy(super.nodePlugins(), MockTransportService.TestPlugin.class);
+    }
+
+    @Before
+    public void setupCluster() throws Exception {
+        internalCluster().ensureAtMostNumDataNodes(0);
+        logger.info("Starting dedicated master node...");
+        internalCluster().startMasterOnlyNode();
+        logger.info("Starting data node...");
+        internalCluster().startNode(onlyRoles(Set.of(DiscoveryNodeRole.DATA_ROLE)));
+        ensureStableCluster();
+    }
+
+    private void ensureStableCluster() {
+        ensureStableCluster(internalCluster().getNodeNames().length, TimeValue.timeValueSeconds(60));
+    }
+
+    public void testNoMlNodesLazyStart() throws Exception {
+        String indexName = "data";
+        client().admin().indices().prepareCreate(indexName).get();
+        client().prepareIndex(indexName)
+            .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+            .setSource("{\"field\": 1, \"other\": 2}", XContentType.JSON)
+            .get();
+
+        DataFrameAnalyticsConfig.Builder dataFrameAnalyticsConfig = new DataFrameAnalyticsConfig
+            .Builder()
+            .setSource(new DataFrameAnalyticsSource(new String[]{indexName}, null, null))
+            .setAnalysis(new OutlierDetection.Builder().build())
+            .setDest(new DataFrameAnalyticsDest("foo", null));
+        {
+            String analyticsId = "not-lazy-dfa";
+            client().execute(
+                PutDataFrameAnalyticsAction.INSTANCE,
+                new PutDataFrameAnalyticsAction.Request(dataFrameAnalyticsConfig.setId(analyticsId).build()))
+                .actionGet();
+            Exception ex = expectThrows(Exception.class,
+                () -> client().execute(
+                    StartDataFrameAnalyticsAction.INSTANCE,
+                    new StartDataFrameAnalyticsAction.Request(analyticsId)
+                ).actionGet());
+            assertThat(ex.getMessage(), containsString("No ML node to run on"));
+        }
+        {
+            String analyticsId = "lazy-dfa";
+            client().execute(
+                PutDataFrameAnalyticsAction.INSTANCE,
+                new PutDataFrameAnalyticsAction.Request(dataFrameAnalyticsConfig.setId(analyticsId).setAllowLazyStart(true).build()))
+                .actionGet();
+            client().execute(StartDataFrameAnalyticsAction.INSTANCE, new StartDataFrameAnalyticsAction.Request(analyticsId)).actionGet();
+            // it is starting lazily
+            assertBusy(() -> {
+                assertThat(client().execute(GetDataFrameAnalyticsStatsAction.INSTANCE,
+                    new GetDataFrameAnalyticsStatsAction.Request(analyticsId))
+                    .actionGet()
+                    .getResponse()
+                    .results()
+                    .get(0)
+                    .getState(), equalTo(DataFrameAnalyticsState.STARTING));
```

Comment:
Could there be a problem that this busy loop won't "catch" the moment in which the state is `STARTING` because analytics will advance to the next state too quickly?

Reply:
no. Because there is no ML node to assign to.